many neutron-ha-tool fixes

aspiers opened the pull request:
**In a Crowbar environment, the retry commit requires https://github.com/crowbar/crowbar-openstack/pull/344 and also https://review.openstack.org/#/c/297664/ in order to be effective.**

Whilst looking into [bsc#965886 (L3: neutron-ha-tool not migrating dhcp networks during failover)](https://bugzilla.suse.com/show_bug.cgi?id=965886), I discovered many instances of braindead code in `neutron-ha-tool`.  In particular, previous authors seemed to think that `LOG.exception()` aborts the code flow, which it doesn't, so several corner cases just completely fail.

Rather than submitting all these changes as individual rpm patches like it's 1995, I've forked the original upstream git repo to this fork, resurrected the Icehouse branch (which was tagged as `eol-icehouse`) and created a new `neutron-ha-tool-maintenance` branch from it which can be subject to the normal code review process.  We can then easily repoint https://build.suse.de/package/show/Devel:Cloud:6:Staging/openstack-neutron at it.  Similarly for Cloud 5.

Commit:
Don't try to log an exception before logger is setup

At this point, we haven't even parsed the CLI arguments in order to
determine logging level, so we haven't set up the logger properly yet.
Even if we had, there is no exception to log.

Commit:
Make migrate_router() raise exceptions on error

LOG.exception() should only be used when an exception was caught.  In
this internal helper function, when something goes wrong, it makes sense
to raise exceptions which can optionally be caught by callers.
Previously it was not bailing correctly if the router failed to be
removed.

Commit:
Invert conditionals in order to reduce indentation

Commit:
Fix handling of cases where no L3 agents are alive

When no L3 agents were alive, the code would erroneously invoke
LOG.exception(), which could log an old exception which was harmless and
already caught and handled gracefully.  It would also still attempt to
perform migrations, which would lead to an IndexError when trying to
randomly pick which alive agent to migrate to:

    Traceback (most recent call last):
      File "/usr/bin/neutron-ha-tool", line 629, in <module>
        run(args)
      File "/usr/bin/neutron-ha-tool", line 122, in run
        l3_agent_migrate(qclient, args.noop, args.now)
      File "/usr/bin/neutron-ha-tool", line 314, in l3_agent_migrate
        target_id = random.choice(agent_alive_list)
      File "/usr/lib64/python2.7/random.py", line 275, in choice
        return seq[int(self.random() * len(seq))]  # raises IndexError if seq is empty
    IndexError: list index out of range

Instead we log an error and then give up on any migrations.

Commit:
Remove superfluous blank lines

Commit:
Count migrations even if they weren't completed

To have a log message at the end with an accurate count of how many
migrations were need, we should count all needed migrations, regardless
of whether we're in dry-run mode or whether some migrations failed.

This also makes future refactoring easier.

Commit:
Log migrations inside migrate_router()

Remove code duplication by logging inside migrate_router().  Also, if
another caller of migrate_router() was introduced, it would accidentally
omit to log the migration.  Moving the log message inside
migrate_router() prevents this.

Commit:
flake8 fixes for 58f12c50605

Commit:
Extract migrate_router_safely() function

This removes some code duplication and helps make l3_agent_migrate() and
l3_agent_evacuate (which are already overgrown) more concise and
readable.

Also remove some spurious 'continue' keywords.

Commit:
Extract migrate_l3_routers_from_agent() function

This removes duplicated code, and helps make l3_agent_migrate() and
l3_agent_evacuate() (which are already overgrown) more concise and
readable.

Commit:
Fix handling of invalid agent to evacuate

When the agent to evacuate could not be found, the code would
erroneously invoke LOG.exception(), which could log an old exception
which was harmless and already caught and handled gracefully.  Worse, it
would also still attempt to retrieve the agent id, which would lead to:

    'NoneType' object has no attribute '__getitem__'

Instead we log an error and then give up on any migrations.

Commit:
Improve comment and logging in migrate_router()

In the evacuation case, migrate_router() can be called on alive agents,
not just dead ones.  Also add a debug message.

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -270,50 +272,39 @@ def l3_agent_migrate(qclient, noop=False, now=False):
-                except:
-                    LOG.exception("There was an error migrating a router")
-                    continue
+    migration_count = 0
```
This is already initialized at https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/1/files#diff-937ed0b9cba9efd2521c8fd1c9e8416fR268

aspiers replied:
Good spot - fixed.

jsuchome commented:
I think those commits replacing LOG.exception could be squashed into one, the reasoning is always the same. But in general LGTM

jsuchome commented:
While you were adding also some older patches to this PR (like the one for password file), could you also add a commit with  https://build.suse.de/package/view_file/Devel:Cloud:6:Staging/openstack-neutron/0001-L3-agent-migration-should-only-choose-enabled-agent.patch?expand=1 ?

aspiers commented:
> I think those commits replacing LOG.exception could be squashed into one, the reasoning is always the same. But in general LGTM

The reasoning is the same, but the damage is different and also some cases are handled in different ways (e.g. `raise` vs. `LOG.error` and `sys.exit(1)`), as explained in the commit messages.  So I think it's better to keep separate.

> While you were adding also some older patches to this PR (like the one for password file), could you also add a commit with https://build.suse.de/package/view_file/Devel:Cloud:6:Staging/openstack-neutron/0001-L3-agent-migration-should-only-choose-enabled-agent.patch?expand=1 ?

Yeah, good catch - thanks I will!

aspiers commented:
> Yeah, good catch - thanks I will!

Ah, I had already moved this into a separate branch which I was going to submit as a separate PR, because I didn't realise we already had a patch for it.  Actually I'll just push it and the password patch into the target branch (`SUSE-Cloud`) since we already know we need them - no point including them in this review.

jsuchome commented:
I've tested with this patch, and although I did got failures of neutron-ha-tool, they are not related to this changes - and yes, I do not see any extra errors coming from incorrect exceptions handling.

+1

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -89,11 +91,19 @@ def run(args):
+    if os.getenv('OS_PASSWORD'):
+        os_password = os.environ['OS_PASSWORD']
+    else:
+        with open('/etc/neutron/os_password') as f:
```
Should we check for the file existence?

aspiers replied:
You mean so that we can abort with a more friendly error message, rather than an exception and a stacktrace?  Sounds reasonable to me.  Or do you mean something else?

vuntz replied:
That's what I meant, yes.

aspiers replied:
Fixed in a new commit.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -263,57 +265,45 @@ def l3_agent_migrate(qclient, noop=False, now=False):
+    if len(agent_alive_list) < 1:
+        LOG.error("There are no l3 agents alive to migrate routers onto - "
+                  "aborting!")
+        sys.exit(1)
```
Why exiting? Why not return and let the rest happen?

aspiers replied:
This is explained in the commit message.  But anyway, what do you mean by "the rest"?  What else can be done if there are no l3 agents alive?

vuntz replied:
dhcp replication? (sorry, didn't read all commit messages)

aspiers replied:
How can dhcp replication work if there are no l3 agents alive?

vuntz replied:
Because dhcp is handled by dhcp agents, not l3 agents? (and it could be that the dhcp-agent service is running, but not the l3-agent)

aspiers replied:
Oh yeah, for some reason I was thinking the l3 agent handled dhcp too.

Whilst your concern is technically valid, it's kind of irrelevant because [dhcp rebalancing is done in a separate invocation of this script](https://github.com/openstack/openstack-resource-agents/blob/master/ocf/neutron-ha-tool#L218).  The only case which would matter is if someone invoked `neutron-ha-tool --l3-agent-migrate --replicate-dhcp` manually, which seems unlikely ... and even then, I wonder how useful working dhcp agents really are if there are no l3 agents.  But I can change it if you really think it's worth it.

vuntz replied:
To be honest, I'd change it regardless of what I pointed out because it's bad style (to me) ;-) But I agree it's not super critical.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -404,9 +379,31 @@ def replicate_dhcp(qclient, noop=False):
+    LOG.info("Querying agent_id=%s for routers to migrate away", agent_id)
+    router_id_list = list_routers_on_l3_agent(qclient, agent_id)
+
+    if not noop:
```
There's one thing you're losing here: I think it was intended to actually have the logging when `noop` is set because it was used to know what the tool would do.

aspiers replied:
Damn, you're right.

aspiers replied:
Actually, no I don't think you are :-)  The `if not noop` was already there before, all I did was move it outside the inner loop.  See 8fee8571acfe03ab14d3c0b8a7ef2dabb1807646 and 20f34b96f29b138775be023dbfc2ece1b3aca9b1.

vuntz replied:
But you also moved the logging bits in some other commits :-)

aspiers replied:
Sure, but I still can't see any problems.  If you're convinced about this, please point out a specific log message which used to appear with `--noop` and doesn't any more.

vuntz replied:
https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/1/files#diff-937ed0b9cba9efd2521c8fd1c9e8416fL207

jsuchome replied:
Hm, it's true that this log line is lost in current version.

Commit:
Die more gracefully if password can't be obtained

If the password isn't provided via OS_PASSWORD environment variable or
in /etc/neutron/os_password, then die gracefully.

aspiers commented:
Rebased against latest "upstream" in the `SUSE-Cloud` fork, since I already pushed the commits there which mirror the patches we already had in the packages: 26a7df46 and 29e97599.

Commit:
Make neutron-ha-tool recovery operations retry (bsc#965886)

If neutron-ha-tool's monitor action fails during a window in which the
database or message bus are down, or still starting up, its consequent
start action will fail.  There's no clean way to make Pacemaker do a
sensible number of retries at staggered intervals, so inside we embed
the retry logic inside the tool itself.  Ideally it would have gone in
the OCF RA, but it's uglier doing it in bash.

https://bugzilla.suse.com/show_bug.cgi?id=965886

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -22,6 +22,7 @@
 import argparse
 import random
 import time
+import retrying
```
Wrong import order.

AbelNavarro replied:
LGTM, tried it and works fine. Why is it wrong?

vuntz replied:
List of imports is supposed to be sorted.

That being said, `retrying` is maybe not in the standard library? If not, then it needs a blank like before the `import`.

AbelNavarro commented on files/default/neutron-ha-tool.py:
```diff
@@ -22,6 +22,7 @@
 import time
+import retrying
 from logging.handlers import SysLogHandler
 from collections import OrderedDict
```
Notice that for cloud5 this import line should be:
`from ordereddict import OrderedDict`

aspiers replied:
Taken care of by https://build.suse.de/package/view_file/Devel:Cloud:5:Staging/openstack-neutron/neutron-ha-tool_python26-ordereddict.patch?expand=1

jsuchome commented:
I've also tested it several times now and it seems to work correctly. Do we still need WIP and Do not merge tags?

vuntz commented:
@jsuchome I think the point about `noop` not doing what it was doing before is still open.

aspiers commented:
@jsuchome and also two other items:
1. some `sys.exit(1)` calls need to be changed into `return` or exceptions which get caught higher up
2. the retry behaviour should be optional with parameter(s), not hardcoded

Commit:
Add metavar for --l3-agent-evacuate

Make the usage text include a more helpful argument name:

    [--l3-agent-evacuate AGENT_ID]

Commit:
Explain what happens if agent comes online during wait period

If the "now" parameter to l3_agent_migrate() is false, and an agent
comes online during the period of waiting, migration will be abandoned.
So document this in the docstring.

Commit:
Require exactly one action to be specified

Output an error if no action is specified.

Also output an error if more than one action is specified, because it's
too complicated to support meaningful semantics for exit codes if
multiple actions are run.  For example, if --l3-agent-migrate and
--replicate-dhcp were both specified to the same invocation, each action
could independently have one of three outcomes:

  - no action was required (everything already healthy)
  - some action was required (router migration or dhcp replication)
  - some error was encountered during the action

That would require choosing script exit codes to meaningfully represent
9 possible different outcomes, just for this combination of two actions.

Instead, we just limit invocation to one action at a time, and then
each action can have easily understandable exit codes associated with
its outcomes.

This should have no impact since nothing currently requires multiple
actions within a single run, and even if it did, it could simply invoke
the tool multiple times, once for each action.

Commit:
Use %d not %s for integer output

Commit:
Make return values and exit codes consistent

Commit:
improve variable names

Clarify when variables are referring to hostnames or agent ids.

aspiers commented:
I think I have now addressed all remaining concerns, and some extra ones:
- Retry is now optional via `--retry`, with `--retry-max-interval` customizable
- At considerable effort which I'm not convinced was worth it, got rid of the `sys.exit` calls. At least it should keep @vuntz happy.
- Better exception handling: don't trap arbitrary exceptions, just `NeutronException`
- Many code cleanups

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -385,36 +439,75 @@ def replicate_dhcp(qclient, noop=False):
+        return True
+    except:
+        LOG.exception("Failed to migrate router=%s from agent=%s "
+                      "to agent=%s", router_id, agent_id, target_id)
```
Would it make sense to also log the message from the exception?

vuntz replied:
Doh, I guess that's already done by `LOG.exception`.

aspiers replied:
Yep :-)

aspiers commented:
Tested on SOC6 and it works, although for some reason I have not managed to reproduce a situation where a retry was actually required.  But manual retries from CLI work, so I can't imagine why they wouldn't from the OCF.

aspiers commented:
(BTW this requires adding `--retry` to the two invocations of the tool in the OCF `start` action)

vuntz commented:
+1

aspiers commented:
Submitted [gerrit#I41c37500 (neutron-ha-tool: make start action retry)](https://review.openstack.org/#/c/297664/)

AbelNavarro commented:
+1
